File: csharp/lib/Property/ImageSensorFunctionProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2023 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Structures used to set the functions used in the sensor.
    /// </summary>
    [MessagePackObject]
    public class ImageSensorFunctionProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "image_sensor_function_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Setting auto exposure function.
        /// </summary>
        [Key("auto_exposure")]
        public bool AutoExposure { get; set; }

        /// <summary>
        /// Setting automatic white balance function.
        /// </summary>
        [Key("auto_white_balance")]
        public bool AutoWhiteBalance { get; set; }

        /// <summary>
        /// Brightness value.
        /// </summary>
        [Key("brightness")]
        public int Brightness { get; set; }

        /// <summary>
        /// ISO sensitivity. (100,200,400,800,1600,...)
        /// </summary>
        [Key("iso_sensitivity")]
        public int IsoSensitivity { get; set; }

        /// <summary>
        /// Time of exposure [100usec].
        /// </summary>
        [Key("exposure_time")]
        public int ExposureTime { get; set; }

        /// <summary>
        /// Exposure metering mode.
        /// </summary>
        [Key("exposure_metering")]
        public string ExposureMetering { get; set; } = string.Empty;

        /// <summary>
        /// Gamma correction value.
        /// </summary>
        [Key("gamma_value")]
        public float GammaValue { get; set; }

        /// <summary>
        /// Gain value.
        /// </summary>
        [Key("gain_value")]
        public int GainValue { get; set; }

        /// <summary>
        /// Hue value.
        /// </summary>
        [Key("hue")]
        public int Hue { get; set; }

        /// <summary>
        /// Saturation value.
        /// </summary>
        [Key("saturation")]
        public int Saturation { get; set; }

        /// <summary>
        /// Sharpness value.
        /// </summary>
        [Key("sharpness")]
        public int Sharpness { get; set; }

        /// <summary>
        /// White balance value.
        /// </summary>
        [Key("white_balance")]
        public int WhiteBalance { get; set; }
    }

    /// <summary>
    /// Structure for acquiring functions supported by Component.
    /// For each function of the image sensor, set the counter corresponding
    /// to Component as a boolean value.
    /// </summary>
    [MessagePackObject]
    public class ImageSensorFunctionSupportedProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "image_sensor_function_supported_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Auto exposure supported.
        /// </summary>
        [Key("auto_exposure_supported")]
        public bool AutoExposureSupported { get; set; }

        /// <summary>
        /// Auto white balance supported.
        /// </summary>
        [Key("auto_white_balance_supported")]
        public bool AutoWhiteBalanceSupported { get; set; }

        /// <summary>
        /// Brightness supported.
        /// </summary>
        [Key("brightness_supported")]
        public bool BrightnessSupported { get; set; }

        /// <summary>
        /// Iso sensitivity supported.
        /// </summary>
        [Key("iso_sensitivity_supported")]
        public bool IsoSensitivitySupported { get; set; }

        /// <summary>
        /// Exposure time supported.
        /// </summary>
        [Key("exposure_time_supported")]
        public bool ExposureTimeSupported { get; set; }

        /// <summary>
        /// Exposure metering supported.
        /// </summary>
        [Key("exposure_metering_supported")]
        public bool ExposureMeteringSupported { get; set; }

        /// <summary>
        /// Gamma value supported.
        /// </summary>
        [Key("gamma_value_supported")]
        public bool GammaValueSupported { get; set; }

        /// <summary>
        /// Gain value supported.
        /// </summary>
        [Key("gain_value_supported")]
        public bool GainValueSupported { get; set; }

        /// <summary>
        /// Hue supported.
        /// </summary>
        [Key("hue_supported")]
        public bool HueSupported { get; set; }

        /// <summary>
        /// Saturation supported.
        /// </summary>
        [Key("saturation_supported")]
        public bool SaturationSupported { get; set; }

        /// <summary>
        /// Sharpness supported.
        /// </summary>
        [Key("sharpness_supported")]
        public bool SharpnessSupported { get; set; }

        /// <summary>
        /// White balance supported.
        /// </summary>
        [Key("white_balance_supported")]
        public bool WhiteBalanceSupported { get; set; }
    }
}

File: java/senscord/src/senscord_properties/ImageSensorFunctionProperty.java
/*
 * SPDX-FileCopyrightText: 2019 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Native;
import com.sun.jna.Structure;
import java.util.Arrays;
import java.util.List;

/**
 * Set the functions used in the sensor.
 * 
 * @see "senscord::ImageSensorFunctionProperty"
 */
public class ImageSensorFunctionProperty extends Structure {
    public byte auto_exposure;
    public byte auto_white_balance;
    public int brightness;
    public int iso_sensitivity;
    public int exposure_time;
    public byte[] exposure_metering;
    public float gamma_value;
    public int gain_value;
    public int hue;
    public int saturation;
    public int sharpness;
    public int white_balance;

    public static class ByReference extends ImageSensorFunctionProperty
            implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("auto_exposure", "auto_white_balance", "brightness", "iso_sensitivity",
                "exposure_time", "exposure_metering", "gamma_value", "gain_value", "hue",
                "saturation", "sharpness", "white_balance");
    }

    /**
     * Exposure metering modes.
     */
    public static interface Metering {
        /**
         * None(Disabled).
         */
        public static final String SENSOCRD_EXPOSURE_METERING_NONE = "note";

        /**
         * Average.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_AVERAGE = "average";

        /**
         * Center weighted.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_CENTER_WEIGHTED = "center_weighted";

        /**
         * Spot.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_SPOT = "spot";

        /**
         * Matrix.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_MATRIX = "matrix";
    }

    public ImageSensorFunctionProperty() {
        super();
        exposure_metering = new byte[64];
    }

    /**
     * Get auto exposure value.
     * 
     * @return auto exposure value
     */
    public byte GetAutoExposure() {
        return auto_exposure;
    }

    /**
     * Get auto white balance value.
     * 
     * @return white balance value
     */
    public byte GetAutoWhiteBalance() {
        return auto_white_balance;
    }

    /**
     * Get Brightness value.
     * 
     * @return brightness value
     */
    public int GetBrightess() {
        return brightness;
    }

    /**
     * Get ISO sensitivity value.
     * 
     * @return ISO sensitivity value.
     */
    public int GetIsoSensitivity() {
        return iso_sensitivity;
    }

    /**
     * Get Exposer time.
     * 
     * @return exposure time
     */
    public int GetExposureTime() {
        return exposure_time;
    }

    /**
     * Get exposure metering.
     * 
     * @return exposure metering
     */
    public String GetExposureMetering() {
        return Native.toString(exposure_metering);
    }

    /**
     * Get gamma value.
     * 
     * @return gamma value
     */
    public float GetGammaValue() {
        return gamma_value;
    }

    /**
     * Get gain value.
     * 
     * @return gain value
     */
    public int GetGainValue() {
        return gain_value;
    }

    /**
     * Get hue value.
     * 
     * @return hue value
     */
    public int GetHue() {
        return hue;
    }

    /**
     * Get saturation value.
     * 
     * @return saturation value
     */
    public int GetSaturation() {
        return saturation;
    }

    /**
     * Get sharpness value.
     * 
     * @return sharpness value
     */
    public int GetSharpness() {
        return sharpness;
    }

    /**
     * Get white balance.
     * 
     * @return white balance
     */
    public int GetWhiteBalance() {
        return white_balance;
    }

    /**
     * Set auto exposure value.
     * 
     * @param auto_exposure aute exposure value
     */
    public void SetAutoExposure(byte auto_exposure) {
        this.auto_exposure = auto_exposure;
    }

    /**
     * Set auto white balance value.
     * 
     * @param auto_white_balance white balance value
     */
    public void SetAutoWhiteBalance(byte auto_white_balance) {
        this.auto_white_balance = auto_white_balance;
    }

    /**
     * Set brightness value.
     * 
     * @param brightness brightness value
     */
    public void SetBrightess(int brightness) {
        this.brightness = brightness;
    }

    /**
     * Set ISO sensitivity value.
     * 
     * @param iso_sensitivity ISO sensitivity value.
     */
    public void SetIsoSensitivity(int iso_sensitivity) {
        this.iso_sensitivity = iso_sensitivity;
    }

    /**
     * Set exposure time.
     * 
     * @param exposure_time exposure time
     */
    public void SetExposureTime(int exposure_time) {
        this.exposure_time = exposure_time;
    }

    /**
     * Set exposure metering value.
     * 
     * @param exposure_metering exposure metering value
     */
    public void SetExposureMetering(String exposure_metering) {
        Arrays.fill(this.exposure_metering, (byte) 0);
        System.arraycopy(exposure_metering.getBytes(), 0, this.exposure_metering, 0,
                exposure_metering.length());
    }

    /**
     * Set gamma value.
     * 
     * @param gamma_value gamma value
     */
    public void SetGammaValue(float gamma_value) {
        this.gamma_value = gamma_value;
    }

    /**
     * Set gain value.
     * 
     * @param gain_value gain value.
     */
    public void SetGainValue(int gain_value) {
        this.gain_value = gain_value;
    }

    /**
     * Set hue value.
     * 
     * @param hue hue value.
     */
    public void SetHue(int hue) {
        this.hue = hue;
    }

    /**
     * Set saturation value.
     * 
     * @param saturation saturation value
     */
    public void SetSaturation(int saturation) {
        this.saturation = saturation;
    }

    /**
     * Set sharpness value.
     * 
     * @param sharpness sharpness value
     */
    public void SetSharpness(int sharpness) {
        this.sharpness = sharpness;
    }

    /**
     * Set white balance value.
     * 
     * @param white_balance white balance value
     */
    public void SetWhiteBalance(int white_balance) {
        this.white_balance = white_balance;
    }
}
